core/src/replay_stage.rs

```diff
@@ -652,8 +653,14 @@ impl ReplayStage {
             let node_keypair = cluster_info.read().unwrap().keypair.clone();
 
             // Send our last few votes along with the new one
+            let mut last_vote = tower.last_vote();
+            let timestamp_interval =
+                slots_per_interval(bank.slots_per_year(), DEFAULT_TIMESTAMP_INTERVAL_MINS);
+            if bank.slot() % timestamp_interval == 1 {
```

Comment:
when is  `slot % slots_per_30_min == 1` ?

Reply:
slot_per_30_min is a huge number (30*60/0.4 =4500)...  we don't vote on every bank or consider voting on every bank, so we could go a **_long_** time between updates.

Probably need to keep the last timestamp in Tower and just do a Duration:: check for 5 minutes or something...

Reply:
> when is `slot % slots_per_30_min == 1` ?

With current slots_per_year, that happens on slot 1, slot 4501, slot 9001, etc...
I chose 1 instead of 0 so that we'd get a timestamp right away on cluster boot.

The reason I went with a specific slot instead of a duration is that it makes it easier to find these votes in the ledger, which we need to do to calculate a particular block time in history. I like your idea of keeping the last timestamp in Tower, however... Do you have any thoughts on how to find the most recent timestamp Votes from the ledger without scanning the entire duration's-worth of slots?

Regarding the specific interval, 30min was proposed in our initial discussion to keep the extra bytes on the wire small. We can certainly drop the interval, if we are okay

Reply:
on wire size: is it really smaller for us to serialize a struct with an Option<timestamp> than a struct with a timestamp?

the thing is, this code doesn't necessarily get to see bank 1 or bank 4501...

I don't have a good idea for how to troll the ledger offhand.  also: the failed votes with bogus timestamps will still be in the ledger...

Reply:
> on wire size: is it really smaller for us to serialize a struct with an Option than a struct with a timestamp?

None case: 1 byte vs 8 bytes?

> the thing is, this code doesn't necessarily get to see bank 1 or bank 4501...

Yeah, I get that.

> I don't have a good idea for how to troll the ledger offhand. also: the failed votes with bogus timestamps will still be in the ledger...
Yep, but we can check statuses.
I'll think on efficient trolling...

Reply:
> > on wire size: is it really smaller for us to serialize a struct with an Option than a struct with a timestamp?
> 
> None case: 1 byte vs 8 bytes?
>

bincode::serialize() is sneaky!
 
> > the thing is, this code doesn't necessarily get to see bank 1 or bank 4501...
> 
> Yeah, I get that.

so the thinking is that *somebody* is gonna hit the 1s?


> > I don't have a good idea for how to troll the ledger offhand. also: the failed votes with bogus timestamps will still be in the ledger...
> 
> Yep, but we can check statuses.
> I'll think on efficient trolling...

if Tower remembered the last vote that had a timestamp, you could do:

```
if last_timestamp_slot % interval < bank.slot() % interval {
}
```
then everybody'd be sending a timestamp every half hour, offset by several slots

Reply:
> so the thinking is that _somebody_ is gonna hit the 1s?

No, you're right that this approach needs to change.

> if Tower remembered the last vote that had a timestamp, you could do:
> 
> ```
> if Tower.last_timestamp_slot % interval < bank.slot() % interval {
> }
> ```
> 
> then everybody'd be sending a timestamp every half hour, offset by several slots

Yes, nice. Just landing on that `<` myself.